### ISSUE TITLE
datastore: Implement "last_updated" for buckets

### ISSUE THREAD

**Issue by johan-bjareholt** [ActivityWatch/aw-server-rust#32]

**Comment by ErikBjare** [comment]
Does this mean I'll finally get my `data` attribute for buckets as well? If so, we should probably come up with an attribute naming convention for metadata with a specific meaning (to the server) and metadata without well-defined meaning. Not sure if this should be the final distinction, but I'm thinking something along the lines of `__` (dunder) in Python and `$` in MongoDB queries.

Like so:

```js
data: {
  "$last_updated": "2019-01-01T12:00Z",
  "name": "A user-friendly name",
  "description": "Describes the content of the bucket in better detail",
}
```

Also, could we get "$last_modification" or something similar that only updates on data edits or event deletions? (i.e. not on simple heartbeat-type event extensions which merely append data to the bucket, in a sense). Would make sense for the caching method I had in mind.

>Will slow down the query of buckets quite significantly as the cache is not enough anymore

What do you mean by this, why would it slow it down?

**Comment by johan-bjareholt** [comment]
> Does this mean I'll finally get my data  attribute for buckets as well? If so, we should probably come up with an  attribute naming convention for metadata with a specific meaning (to  the server) and metadata without well-defined meaning. Not sure if this  should be the final distinction, but I'm thinking something along the  lines of __ (dunder) in Python and $ in MongoDB queries.

No it doesn't unfortunately.

I don't think we should use prefixes but a sub-object "data" or "metadata" instead to make it very clear that all fields in there are not required.

But last_updated is different though as it's not really a set value, but just a computed value which is never saved. What to do about that though I'm not sure, but as I said I'm not a fan of undescriptive prefixes so maybe a "computed" object for something like that instead?.

> In the longer run, I think it would make sense to move things like hostname and client into the bucket data. What do you think?

I'd be happy to move "name" and "description" into that data/metadata section.
About moving "hostname" in there though would make it optional which I'm not really sure about, it's an important identifier for tracking what host the bucket is owned by. The fact that aw-watcher-web is not able to get the hostname is the bug, not the field itself (unfortunately an unfixable bug though). We could in the future when we have sync support maybe just let aw-server decide the hostname and break the support for the "remote aw-server" functionality.

> Also, could we get "$last_modification" or something similar that only updates on data edits or event deletions? (i.e. not on simple heartbeat-type event extensions which merely append data to the bucket, in a sense). Would make sense for the caching method I had in mind.

last_updated is a computed value, last_modification would need to keep state between restarts so it would have to be written down in the database. I think it would make sense to have that in the "computed" object still though even though it's technically not computed. Suggestions on a better name than "computed" would be appreciated.

> What do you mean by this, why would it slow it down?

It would slow down because we would need another SQL query for each bucket (currently all buckets are cached in RAM). In case of last_modified it would even need a write as well and would be even slower.

**Comment by johan-bjareholt** [comment]
I'll start with adding the "data" section.

**Comment by ErikBjare** [comment]
>About moving "hostname" in there though would make it optional which I'm not really sure about, it's an important identifier for tracking what host the bucket is owned by. The fact that aw-watcher-web is not able to get the hostname is the bug, not the field itself (unfortunately an unfixable bug though). We could in the future when we have sync support maybe just let aw-server decide the hostname and break the support for the "remote aw-server" functionality.

The hostname is not actually a good identifier for devices, since it's not unique. Syncthing does this right (I think, not sure about specifics) it generates a device id from a key pair it has, such that the device (or actually, installation) can always be authenticated and known to be unique.

Everything else sounds good to me, I'm mostly just happy about the ability to add data to buckets themselves (so we can get good metadata for containers).

The idea of `computed` sounds best to me, but I dislike the name. Maybe just `cache`? (since that's essentially what a `computed` would be).

**Comment by johan-bjareholt** [comment]
> The hostname is not actually a good identifier for devices, since it's not unique. Syncthing does this right (I think, not sure about specifics) it generates a device id from a key pair it has, such that the device (or actually, installation) can always be authenticated and known to be unique.

Agreed, but that unique identifier has to be decided at a central point so it wouldn't work for each watcher. We'd have to do as I suggested with aw-server assuming that all watchers which connect to it is on the same host and make it decide what the hostname/uid should be.

**Comment by johan-bjareholt** [comment]
Last updated really is a stupid name now that I think about it, I want to change it to last_event_timestamp.
And then it would also be useful with a first_event_timestamp.

After those two changes we can consider what the usefulness of the "created" property is.

**Comment by johan-bjareholt** [comment]
@ErikBjare Also, what do you think about just calling computed/cached for "metadata"? I think that's the best option.

**Comment by johan-bjareholt** [comment]
Started working on it, made it like this for now

```
"metadata": {
    "first_starttime": "2018-01-01"
    "last_endtime": "2018-01-02"
}
```

**Comment by johan-bjareholt** [comment]
Solved by #32
